controllers/update_flow_steps.go

```diff
@@ -0,0 +1,312 @@
+	ytsaurus.LogUpdate(ctx, fmt.Sprintf("Update flow starting with %s, updating components: %v, all components: %v", ytsaurus.GetUpdateState(), updatingComponents, allComponents))
+	return tree.execute(ctx, ytsaurus, componentManager)
+}
+
+func convertToYtComponents(components []components.Component) []ytv1.Component {
+	result := make([]ytv1.Component, len(components))
+	for i, c := range components {
+		result[i] = ytv1.Component{
+			ComponentType: c.GetType(),
+		}
+	}
+	return result
+}
+
+var terminateTransitions = map[ytv1.UpdateState]ytv1.ClusterState{
+	ytv1.UpdateStateImpossibleToStart: ytv1.ClusterStateCancelUpdate,
+}
+
+type flowCondition func(ctx context.Context, ytsaurus *apiProxy.Ytsaurus, componentManager *ComponentManager) stepResultMark
+
+func flowCheckStatusCondition(conditionName string) flowCondition {
+	return func(ctx context.Context, ytsaurus *apiProxy.Ytsaurus, componentManager *ComponentManager) stepResultMark {
+		if ytsaurus.IsUpdateStatusConditionTrue(conditionName) {
+			return stepResultMarkHappy
+		}
+		return stepResultMarkUnsatisfied
+	}
+}
+
+type flowStep struct {
+	updateState ytv1.UpdateState
+	// For most of the steps, there will be only one next step,
+	// but for some outcome is based on condition result.
+	nextSteps map[stepResultMark]*flowStep
+}
+
+func newSimpleStep(updateState ytv1.UpdateState) *flowStep {
+	return &flowStep{
+		updateState: updateState,
+	}
+}
+
+func newConditionalForkStep(updateState ytv1.UpdateState, unhappyNext *flowStep) *flowStep {
+	return &flowStep{
+		updateState: updateState,
+		nextSteps: map[stepResultMark]*flowStep{
+			stepResultMarkUnhappy: unhappyNext,
+		},
+	}
+}
+
+func (s *flowStep) checkCondition(ctx context.Context, ytsaurus *apiProxy.Ytsaurus, componentManager *ComponentManager) stepResultMark {
+	condition := flowConditions[s.updateState]
+	return condition(ctx, ytsaurus, componentManager)
+}
+
+func (s *flowStep) next(mark stepResultMark) *flowStep {
+	if mark == stepResultMarkUnsatisfied {
+		return nil
+	}
+	return s.nextSteps[mark]
+}
+
+func (s *flowStep) addHappyPath(next *flowStep) *flowStep {
+	if s.nextSteps == nil {
+		s.nextSteps = make(map[stepResultMark]*flowStep)
+	}
+	s.nextSteps[stepResultMarkHappy] = next
+	return next
+}
+
+func (s *flowStep) chain(steps ...*flowStep) *flowStep {
+	current := s
+	for _, step := range steps {
+		current.addHappyPath(step)
+		current = step
+	}
+	return current
+}
+
+// for the fast lookup of the step by the update state
+// otherwise we can traverse the graph each time
+type flowTree struct {
+	index map[ytv1.UpdateState]*flowStep
+	head  *flowStep
+	tail  *flowStep
+}
+
+func newFlowTree(head *flowStep) *flowTree {
+	return &flowTree{
+		index: map[ytv1.UpdateState]*flowStep{
+			head.updateState: head,
+		},
+		head: head,
+		tail: head,
+	}
+}
+
+func (f *flowTree) execute(ctx context.Context, ytsaurus *apiProxy.Ytsaurus, componentManager *ComponentManager) (bool, error) {
```

Comment:
I think it would be benefitial to make bool return param named or explain when it is true or false in the comment. Otherwise, you need to check code to understand the sematics.